CIP-????/README.md

```diff
@@ -0,0 +1,530 @@
+   ; 2) every transaction_index must be strictly smaller than the
+   ;    length of transaction_bodies
+```
+
+Just like with the transaction cddl, the block's transaction bodies and the transaction witness sets
+will be merged, and the constraints must be updated accordingly:
+
+```cddl
+; This is re-posted here for reference.
+transaction_piece =
+  { 0: transaction_body
+  , 1: transaction_witness_set ; The witness set for this transaction body.
+  }
+
+; A specific transaction in the block is a specific batch of transaction pieces. 
+block_transaction = [* transaction_piece]
+
+block =
+  [ header
+  , block_transactions   : [* block_transaction] ; This is new
+  , auxiliary_data_set   : {* transaction_index => auxiliary_data }
+  , invalid_transactions : [* transaction_index ]
+  ]; Valid blocks must also satisfy the following constraint:
+   ; 1) every transaction_index must be strictly smaller than the
+   ;    length of block_transactions
+```
+
+### Transaction Size Limit
+
+A transaction broken into pieces should still have approximately the same total size as the whole
+transaction. There are a few things to note:
+
+- If two transaction pieces require the same witness (ie, both pieces require a signature from
+Alice), this witness would now effectively appear twice.
+- Certain items can appear in multiple pieces (eg, the same input can be spent/referenced, tokens can
+be minted/burned that would otherwise cancel out, etc).
+
+Despite the above, it is probably fine to leave the maximum transaction size as is, at least to
+start. The transaction fee would naturally incentivize users/aggregators to minimize the amount of
+duplication in a transaction.
+
+### Block Size Limit
+
+Due to the possible duplicate items in each transaction and the fact that a block is made up of
+multiple transactions, the issues in the Transaction Size Limit section are slightly exacerbated.
+However, it should still be fine to keep the current block size limit, at least initially. Just like
+with transactions, the transaction fee would naturally incentivize users to minimize duplications.
+
+### Smart Contract Execution
+
+While the cryptographic keys only need to sign the relevant transaction pieces, the smart contracts
+would be executed against the full transaction. This requirement enables the current smart contract
+interface to remain mostly untouched. The flow is essentially:
+
+1. Deterministically derive the whole transaction from the list of pieces.
+2. Generate the script context for the whole transaction.
+3. Evaluate the smart contract using this script context.
+
+When combining the transaction pieces, the required executions (ie, script + datum + redeemer) can
+be built up as well. Duplicate executions should be removed. For example, if two transaction pieces
+look to spend the same smart contract input using the same redeemer, there is no reason to evaluate
+this scenario twice since they are guaranteed to have the same result.
```

Comment:
> How can you spend the same input twice?

You wouldn't. You would remove duplicates when merging the pieces. If you have a list `[1,2,3,4]` and another `[2,5,6]` merging them (and removing duplicates) would give `[1,2,3,4,5,6]`. So input `2` is *listed* twice, but only *spent* once.

Smart contracts are a definite weak-point of this CIP, but the new Validation Zones revision seems to address this weak-point while still incorporating the main ideas from this CIP. I personally prefer the new Validation Zones CIP over my own for this reason.

Reply:
> You would remove duplicates when merging the pieces

So, as I said you can't remove duplicates because of smart contracts.

And with validation zones you can't have duplicates.